File: services/downloader.js
/**
 * Get the file associated with given message_id 
 * and start downloading it.
 * File could be split in more than one file, so
 * we have to get the right piece of file and download via `range request`.
 * User can choose to download 'entire original file' or 'single part' or 'a portion of file'.
 * File parts could be located in different channel. 
 * Keep in mind: user must have full access of channel in order to download files
 * 
 * Example of input data:
 * 
 * type FilePartData = {
 *   ch: string;    // telegram channel_id which contains file part (last 10 numbers, absolute number)
 *   msg: number;   // telegram message_id associated to the file part
 *   size: number   // size of the file part in bytes
 * }
 * 
 * const DATA : FilePartData[] = [
 *   {
 *     ch: 'xxxxx',        
 *     msg: 1,             
 *     size: 459578337
 *   }, {
 *     ch: 'xxxxx',
 *     msg: 2,
 *     size: 452920795
 *   }, {
 *     ch: 'yyyyy',
 *     msg: 3,
 *     size: 4529207
 *   }, {
 *     ch: 'yyyyyy',
 *     msg: 4,
 *     size: 4529207
 *   }
 * ];
 * 
 */

const { UPLOAD_CHUNK } = require('../config');
const Logger = require('../logger');

const Log = new Logger('Downloader');

class Downloader {

  /* if true stops downloading */
  aborted = false;
  
  /* total size of file */
  totalsize = 0;

  /* range of bytes to be downloaded */
  range = null;

  client = null;

  /**
   * constructor
   * @param {string} id: the unique identifier of task
   * @param {FilePartData[]} data: data to be processed (FilePartData: see above comments)
   * @param {number} start: the start range
   * @param {number} end: the end range
   */
  constructor(client, data, start, end) {
    this.data = data;
    this.range = {
      start, end
    };
    this.totalsize = this.data.reduce((acc, curr) => acc + curr.size, 0);
    this.client = client;
  }


  get Range() {
    return {
      ...this.range,
      totalsize: this.totalsize
    }
  }

  /**
   * Stop downloading
   */
  stop() {
    Log.warn('aborted');
    this.aborted = true;
  }

  /**
   * @param {Stream} destination: the destination stream
   */
  async execute(destination) {

    const files = [];

    let currentSizePosition = 0;
    let currentIndex = 0;

    // Calculate the full range stack to be downloaded
    while( true ) {

      const file = this.data[currentIndex];
      let fileToAdd = null;

      if ( this.range.start < (currentSizePosition + file.size) ) {
        // found first chunk of file part to add to download queue
        fileToAdd = {
          index: currentIndex,
          file,
          start:  files.length == 0 ? this.range.start - currentSizePosition : 0
        };

        if ( this.range.end <= (currentSizePosition + file.size) ) {
          
          fileToAdd.end = file.size - ( (currentSizePosition + file.size) - this.range.end ) + 1;
          files.push(fileToAdd);
          break;

        } else {
          
          fileToAdd.end = file.size;
          files.push(fileToAdd);

        }
      }

      currentSizePosition += file.size;

      currentIndex++;
    }

    destination.setDefaultHighWaterMark(false, UPLOAD_CHUNK * 2);
    destination.resume();
    let tgChannel = null;

    for await (const item of files) {

      const {start, end, file} = item;
      const {ch, msg} = file;
      
      if ( !tgChannel || tgChannel.id !== ch) {
        Log.debug('getting channel', ch);
        tgChannel = await this.client.getChannel( ch );
      }
  
      Log.debug('getting message', msg);
      const message = await this.client.getMessage({id: tgChannel.id, hash: tgChannel.access_hash}, msg);
      const {media} = message;
      const {document} = media;
      const {id, access_hash, file_reference} = document;
  
      Log.debug('ready for download, range:', `${start}-${end}`);
      await this.performStream({id, access_hash, file_reference}, start, end, destination);

      if ( this.aborted ) {
        break;
      }
  
    }

    // close stream
    destination.end();
  }


  /**
   * 
   * @param {string} id: the telegram file_id
   * @param {number} start: the start range
   * @param {number} end: the end range 
   * @param {Stream} stream: destination stream
   */
  async performStream({id, access_hash, file_reference}, start, end, stream) {

    // telegram chunk (1MB)
    const CHUNK = 1 * 1024 * 1024;
  
    // calculate the start offset of download
    let offset = start - (start % CHUNK);
    let first = true;
  
    let needStop = false;

    Log.debug('stream from', start, end);
  
    while(true) {
      // Log.debug(`[${this.id}]`, 'get file from telegram', offset);
      const tgFile = await this.client.getFile({id, access_hash, file_reference}, offset, CHUNK);
  
      let firstByte = 0;
      let lastByte = CHUNK;
  
      if ( first ) {
        firstByte = start - offset;
        first = false;
      }
  
      if ( offset + tgFile.bytes.length >= end ) {
        lastByte = end - offset;
        needStop = true
      }
  
      const buf = Uint8Array.prototype.slice.call(tgFile.bytes, firstByte, lastByte);
      // Log.debug(`[${this.id}]`, 'write', buf.length);
      stream.write( buf );
  
      offset += CHUNK;
  
      if ( needStop || this.aborted ) { 
        break
      }
    }
  
  }

}



module.exports = Downloader;

File: python/services/downloader.py
from services.telegram import TelegramApi
from services.database import TGFile
from configuration import UPLOAD_CHUNK
import logging

Log = logging.getLogger('Downloader')

PART_TO_LOG_DEBUG = 100 * 1024 * 1024
PART_TO_LOG_INFO = 500 * 1024 * 1024

class Downloader():
  
  aborted = False

  client = None
  channel_id = None
  parts = None
  range_start = 0
  range_end = -1
  totalsize = 0
  file = None

  def __init__(self, client: TelegramApi, file: TGFile, start: int, end: int):
    self.client = client
    self.channel_id = file.channel
    self.file = file
    self.parts = file.parts
    self.range_start = start
    self.range_end = end

    self.totalsize = 0

    for part in file.parts:
      self.totalsize += part.size
    
  
  def stop(self):
    self.aborted = True
  

  async def execute(self, destination, awaited = True):

    files = []

    currentSizePosition = 0
    currentIndex = 0

    if self.file and self.file.content is not None and self.file.content_length():
      if awaited:
        await destination.write( file.content )
      else:
        destination.write( file.content )
      
      try:
        if awaited:
          await destination.write_eof()
        else:
          destination.write_eof()
      except Exception:
        Log.warn(f"cannot write_eof")
      
      return

    # Calculate the full range stack to be downloaded
    while( True ):

      file = self.parts[currentIndex]
      fileToAdd = None

      if ( self.range_start < (currentSizePosition + file.size) ):
        # found first chunk of file part to add to download queue
        fileToAdd = {
          'index': currentIndex,
          'file': file,
          'start': self.range_start - currentSizePosition if len(files) == 0 else 0
        }

        if ( self.range_end <= (currentSizePosition + file.size) ):
          
          fileToAdd['end'] = file.size - ( (currentSizePosition + file.size) - self.range_end ) + 1
          files.append(fileToAdd)
          break

        else:
          fileToAdd['end'] = file.size
          files.append(fileToAdd)
      
      currentSizePosition += file.size
      currentIndex += 1
    
    
    for item in files:

      start = item['start']
      end = item['end']
      file = item['file']
      msg = file.messageid
      
      Log.debug(f"getting message {msg}")
      message = await self.client.get_message(self.channel_id, msg)
      media = TelegramApi.get_media_from_message( message )
  
      Log.debug(f"ready for download, range: {start}-${end}")

      await self.perform_stream(media.filedata.media_id, media.filedata.access_hash, media.filedata.file_reference, media.filedata.dc_id, start, end, destination, awaited)

      if ( self.aborted ):
        break
    
    try:
      await destination.write_eof()    
    except Exception:
      Log.warn(f"cannot write_eof")

  
  async def perform_stream(self, id, hash, reference, dc, start, end, destination, awaited = True):
    # telegram chunk (1MB)
    CHUNK = UPLOAD_CHUNK * 2 # 1 Mb
  
    # calculate the start offset of download
    offset = start - (start % CHUNK)
    first = True
  
    needStop = False

    total_file_downloaded = 0

    offset = int(offset)

    Log.debug(f"stream from {start} to {end}, starting from {offset}")

    while (True):
      tgFile = await self.client.get_file(id, hash, reference, offset=offset, limit=CHUNK, dc=dc)
  
      firstByte = 0
      lastByte = CHUNK
  
      if ( first ):
        firstByte = start - offset
        first = False
  
      if ( offset + len(tgFile.bytes) >= end ):
        lastByte = end - offset
        needStop = True
      
      firstByte = int(firstByte)
      lastByte = int(lastByte)
  
      buf = tgFile.bytes[firstByte : lastByte]
      # Log.debug(f"send buffer: from {offset} ({len(buf)} bytes)")
      if awaited:
        resp = await destination.write( buf )
      else:
        resp = destination.write( buf )
      
      total_file_downloaded += len(buf)

      if total_file_downloaded % PART_TO_LOG_DEBUG == 0:
        Log.debug(f"downloaded {total_file_downloaded} bytes of '{self.file.filename}'")
      if total_file_downloaded % PART_TO_LOG_INFO == 0:
        Log.info(f"downloaded {total_file_downloaded} bytes of '{self.file.filename}'")

      #Log.debug(f"wrote {len(buf) == resp}, {len(buf)} bytes")

      offset += CHUNK
  
      if ( needStop or self.aborted ):
        break